src/Grpc.Net.Client/Internal/Retry/HedgingCall.cs

```diff
@@ -0,0 +1,409 @@
+        {
+            _hedgingPolicy = hedgingPolicy;
+            _activeCalls = new List<IGrpcCall<TRequest, TResponse>>();
+
+            if (_hedgingPolicy.HedgingDelay > TimeSpan.Zero)
+            {
+                _delayInterruptTcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                _hedgingDelayCts = new CancellationTokenSource();
+            }
+        }
+
+        private async Task StartCall(Action<GrpcCall<TRequest, TResponse>> startCallFunc)
+        {
+            GrpcCall<TRequest, TResponse> call;
+            lock (Lock)
+            {
+                if (CommitedCallTask.IsCompletedSuccessfully())
+                {
+                    // Call has already been commited. This could happen if written messages exceed
+                    // buffer limits, which causes the call to immediately become commited and to clear buffers.
+                    return;
+                }
+
+                OnStartingAttempt();
+
+                call = HttpClientCallInvoker.CreateGrpcCall<TRequest, TResponse>(Channel, Method, Options, AttemptCount);
+                _activeCalls.Add(call);
+
+                startCallFunc(call);
+
+                SetNewActiveCallUnsynchronized(call);
+            }
+
+            Status? responseStatus;
+
+            HttpResponseMessage? httpResponse = null;
+            try
+            {
+                call.CancellationToken.ThrowIfCancellationRequested();
+
+                CompatibilityExtensions.Assert(call._httpResponseTask != null, "Request should have be made if call is not preemptively cancelled.");
+                httpResponse = await call._httpResponseTask.ConfigureAwait(false);
+
+                responseStatus = GrpcCall.ValidateHeaders(httpResponse, out _);
+            }
+            catch (Exception ex)
+            {
+                call.ResolveException(GrpcCall<TRequest, TResponse>.ErrorStartingCallMessage, ex, out responseStatus, out _);
+            }
+
+            // Check to see the response returned from the server makes the call commited
+            // Null status code indicates the headers were valid and a "Response-Headers" response
+            // was received from the server.
+            // https://github.com/grpc/proposal/blob/master/A6-client-retries.md#when-retries-are-valid
+            if (responseStatus == null)
+            {
+                // Headers were returned. We're commited.
+                CommitCall(call, CommitReason.ResponseHeadersReceived);
+
+                // Wait until the call has finished and then check its status code
+                // to update retry throttling tokens.
+                var status = await call.CallTask.ConfigureAwait(false);
+                if (status.StatusCode == StatusCode.OK)
+                {
+                    // Success. Exit retry loop.
+                    Channel.RetryThrottling?.CallSuccess();
```

Comment:
Lot's of nullable checks for RetryThrottling, anyway we can make it non-nullable?

Reply:
Still nullable, but I've centralized calls to retry throttling and the number of null checks.